### PR TITLE
WIP: Implement failedSave Event

### PR DESCRIPTION
**Pull request by cleptric**
This is WIP for implementing the requested `failedSaved` and `failedDelete` events in #9653
It turned out that this isn't as trivial as I thought, so I wanted to collect some feedback before I go on.
I started with `failedSave`.

Of course I'll add way more tests when I proceed.

### CODE REVIEW

**Comment by dereuromark** [comment]
This all looks to me rather non-straightforward - compared to the easy exception solution I have been using so far. My idea was to have a simple direct feedback for the non interactive DB save operations that would otherwise silently stay erroring undetected. This, and using the events system, doesn't quite yet seem like the way to go.

**Comment by cleptric** [comment]
Implementing the ˋTable::failedSaveˋ in on of your Tables, or even an ˋAppTableˋ would give the users cpmplete freedom to implement custom logic in case of a failed save. I guess this is a good compromise when summarizing all the different opinions in the linked issue.

**Comment by dereuromark** [comment]
How can you decide in AppTable callback etc if the current fail is Inside an interactive validation form scenario or just a non-interactive save? Maybe I am missing something.

File: src/ORM/Table.php [hunk 1]
```diff
@@ -1479,6 +1484,8 @@ public function save(EntityInterface $entity, $options = [])
         ]);
 
         if ($entity->errors()) {
+            $this->dispatchEvent('Model.failedSave', compact('entity', 'options'));
```
Comment (robertpustulka) [inline src/ORM/Table.php]: I'm not sure we should do a check this early. This one means probably that validation failed. And this failure should be handled in client code or using `Marshaller` event-cycle.

**Comment by robertpustulka** [comment]
For me `Model.failureSave` should be fired *after* and only *if* `Model.beforeSave` has fired and instead of `Model.afterSave`.

For failed rules we could also add an event inside rules event-cycle.

Note this diagram: 
image

I think there should be events fired for those failures that escape the save process.

**Comment by cleptric** [comment]
@robertpustulka This sounds pretty reasonable to me :)

@dereuromark This is of course a valid concern. I guess that I would pass an option, which I then check for in the callback.

Correct me if I'm wrong, but I got the impression, that throwing an exception is not the way to go here. So are there any alternatives we haven't discussed before?
Working with the event system was always a big part of writing code for my cake apps, so for me, that doesn't feels so clunky.

**Comment by robertpustulka** [comment]
For @dereuromark 's original idea throwing an exception would be a best solution, but I think that the idea has evolved into enhancing the event cycle. For me is a more flexible approach. I think that CakePHP was really missing some way to observe save/delete failures. Sometimes there's some work being done in beforeSave and failureSave is a good place to clean it up if something goes wrong.

**Comment by dereuromark** [comment]
> Correct me if I'm wrong, but I got the impression, that throwing an exception is not the way to go here

@cleptric Why do you think that? How else could you "halt" execution in that point in time to avoid further issues along the road from this failure of saving?
Just logging here is usually not quite sufficient.
Let's say I have a cronjob that saves stuff, and then continues to work on related data.
I would like to bail early here of course if the first save already didnt work. I also need to be notified via error logs.
What I do not want however is that a callback in the table triggers all save fails to convert into logs/exceptions as those might be just because of validation/domain rule failure inside a form.

If we can separate those two scenarios with the above approach I am all in :) Right now I do not quite see it, though. And manually wrapping every save with if () was what my solution so far prevents.

**Comment by lorenzo** [comment]
I don't like this idea, mainly because I'm not sure what use cases this is trying to solve.

So far, it seem like the only likely usage of this feature is to log whether or not an entity passed validation, but that does not seem overly useful to me.

**Comment by lorenzo** [comment]
To people contributing in this thread... What are the real world usages in which you see yourself using this callback?

**Comment by robertpustulka** [comment]
@lorenzo what @dereuromark said: halting code execution in cron jobs. Doing it through events seems like a morew flexible, extendable and observer-like way. Also I had some situation where I'd done some changes to entity on before save and I had to revert that changes when a save failed. Doing it in a same listener would be a cleaner way to handle such situations.

**Comment by lorenzo** [comment]
> Also I had some situation where I'd done some changes to entity on before save and I had to revert that changes when a save failed.

That sounds like you should `clone` the entity before saving

If the only purpose of the feature is to have your app break when a save fails, then this is not a great way, since you still have to implement event listeners to make your application break. In that case exceptions are a much better solution.

The problem with exceptions is that no one has done the work of *thinking* what would happen if we make it behave that way in all scenarios.

I think the best alternative is to introduce a new method `saveStrict()` in the table. With this implementation:

```php
public function saveStrict($entity, $options = [])
{
   $saved = $this->save($entity, $options);
   if ($saved === false) {
       throw SomeException("Message", $entity);
   }

   return $saved;
}
```

**Comment by lorenzo** [comment]
In the other had, it was so easy to implement that makes you wonder why people are not doing this already

**Comment by dereuromark** [comment]
@lorenzo I am doing this already. A new method or a strict option is essentially the same thing.
See my original issue and the linked code.
The question was why such a "best practice" approach isn't simply added to the core.
Especially since often times overwriting plugin code extending the core Table is not that trivial for app development. So having this centralized in the core in a proper way might make sense.

**Comment by cleptric** [comment]
@lorenzo @dereuromark All my thoughts on this topic are based on this issuse: https://github.com/cakephp/cakephp/issues/9653#issuecomment-276689062
I just summarised the different opinions and came up with a first idea :)

Personally, I really like the new `failedSave` event and also, a `saveStrict` Method would be nice to have, too.

**Comment by lorenzo** [comment]
@dereuromark The problem with the option is that you would expect this option to be propagated down to associations, and I think that is prone to quite a few kinds of problems down the road.

The strict method can be added to the core, I have no problem with that 😃

**Comment by cleptric** [comment]
Ok, then I'll do a second PR and introduce a `saveStrict` Method. We then can continue here to discuss if we want/need to implement the new events.
Does this sound good to you?

**Comment by lorenzo** [comment]
@cleptric sounds good

**Comment by dereuromark** [comment]
Interesting. I wonder what issues arise from the passing it to associations.

What about the 2nd part of my ticket and linked code, deleting?
https://github.com/dereuromark/cakephp-shim/commit/d5393104d183668d2743a9ec8008b48466df0942
Do we have the same agreement here, then?

**Comment by lorenzo** [comment]
having it as an option in delete is less risky than with save, but if we add a `saveStrict` function, then probably we want a `deleteStrict` for symmetry.

**Comment by dereuromark** [comment]
Is saveOrFail() etc a better name maybe?

**Comment by ADmad** [comment]
Yeah `saveOrFail()` would be inline with existing `findOrFail()`

**Comment by robertpustulka** [comment]
+ `deleteOrFail` and we're home.

**Comment by ADmad** [comment]
Ding! Ding! Ding!

**Comment by robertpustulka** [comment]
Hmm, maybe `newEntityOrFail()` for validation failures? :smile:

**Comment by dereuromark** [comment]
Let's tackle it in small pieces :) I don't think for validation it makes a lot of sense, as those will be passed to a save() call anyway, especially in those programmatic use cases mentioned above.

**Comment by lorenzo** [comment]
`newEntityOrGTFO`

**Comment by markstory** [comment]
Closing in preference of #10164 which is a simpler solution, and simpler to use.

**Pull request closed without merging**